### DISCUSSION TITLE
How to reuse AssertJ assertions on custom assertions?

### DISCUSSION THREAD

**Discussion by xtaixe** [assertj/assertj#2679] [Q&A]
The examples on the custom assertions docs implement the assertion logic directly, but for more complex checks it would be nice to reuse the assertions that AssertJ already provides and I'm not sure how to that properly.

First of all, I'm not sure if classes in the `internal` package (e.g. `Iterables.instance()` or `Strings.instance()`) are intended to use outside AssertJ and could help somehow.

Other than that after looking around I think something like this should be done to respect the error, description, etc. if set by the user:
```java
public S customComplexAssertionForInternalProperty() {
    var assertion = assertThat(property)
            .withFailMessage("Default message")
            .withRepresentation(getWritableAssertionInfo().representation())
            .describedAs(getWritableAssertionInfo().description());

    if (getWritableAssertionInfo().overridingErrorMessage() != null) {
        assertion.withFailMessage(getWritableAssertionInfo().overridingErrorMessage())
    }

    assertion.complexAssertion()
    return myself;
}
```

I also looked into the `withAssertionState` method, but although it partly fits the bill and would fill even more information depending on the assertion class, it's currently package private and there would still be the need to check somehow if the user set their own values or appropriate defaults should be used (e.g. `"Default message"` vs `getWritableAssertionInfo().overridingErrorMessage()` above).

Would it be possible to add something to more easily support this? Or is there already a recommended way or reusing assertions that could be documented? Any guidance will be appreciated.

Thanks.

**Comment by scordio** [comment]
> for more complex checks it would be nice to reuse the assertions that AssertJ already provides

Assuming you are creating a custom assertion class, you might want to extend `AbstractAssert` or `AbstractObjectAssert`:
* `AbstractAssert` exposes a smaller API surface and you would have full control over the assertion methods you'd like to add on top of the common ones (`isEqualTo`, `isInstanceOf`, `isNull`, etc.).
* `AbstractObjectAssert` offers some additional methods out of the box that might be handy for asserting instance properties (`extracting`, `returns`, `doesNotReturn`)

> I'm not sure if classes in the `internal` package (e.g. `Iterables.instance()` or `Strings.instance()`) are intended to use outside AssertJ and could help somehow.

They're not intended to be used outside AssertJ and you would get a failure in both JPMS and OSGi world if you try to do that.

Could you provide some more context on what you're trying to achieve? Would you have the chance to sketch the custom assertion you are trying to design in a GitHub project, together with the type you are targeting?

**Reply by xtaixe** [reply]
Inheritance won't work on all cases, since in some cases it won't make sense to inherit and in others if you want to reuse assertions from 2 different classes you won't be able to inherit from both.

As an example of what we want to do, let's say I have the following class:
```java
public interface CommandResult {
    Optional<Integer> getExitCode();
    List<String> getStdOut(); // stdout lines
    List<String> getStdErr(); // stderr lines
}
```

And I want to write a `CommandResultAssert` class, that has assertions like the following:
```java
public S stdOutAllSatisfy(Consumer<String> requirements) {
    // reuse allSatisfy AssertJ assertion for stdout lines list
    return myself;
}

public S stdErrAllSatisfy(Consumer<String> requirements) {
    // reuse allSatisfy AssertJ assertion for stderr lines list
    return myself;
}
```

**Reply by xtaixe** [reply]
@scordio any further thoughts on this?

**Reply by scordio** [reply]
Yes, I'd like to create a small example for your use case. I will try to spend some time on it later in the day.

**Reply by scordio** [reply]
@xtaixe apologies for the delay.

Here is my basic proposal with purely navigation methods:

```java
public class CommandResultAssert extends AbstractObjectAssert<CommandResultAssert, CommandResult> {

  public CommandResultAssert(CommandResult commandResult) {
    super(commandResult, CommandResultAssert.class);
  }

  public OptionalAssert<Integer> exitCode() {
    isNotNull();
    return assertThat(actual.getExitCode());
  }

  public ListAssert<String> stdOut() {
    isNotNull();
    return assertThat(actual.getStdOut());
  }

  public ListAssert<String> stdErr() {
    isNotNull();
    return assertThat(actual.getStdErr());
  }

}
```

Having also:

```java
public class Assertions extends org.assertj.core.api.Assertions {

  public static CommandResultAssert assertThat(CommandResult actual) {
    return new CommandResultAssert(actual);
  }

}
```

You could write:

```java
assertThat(commandResult).exitCode().hasValue(1);
assertThat(commandResult).stdOut().allSatisfy(line -> assertThat(line).isNotBlank());
assertThat(commandResult).stdErr().isEmpty();
```

**Reply by xtaixe** [reply]
That is an option. The drawback here is that it removes the ability to:

1. Combine assertions and still reuse AssertJ assertions internally, e.g. `successfulStdOutSatisfies` would mean `exitCode == 0 && stdOutSatisfiesAssertion`. This might not be a very good example, but I can imagine domains where a combined assertion makes much more sense to the user that having to extract individual field assertions.
2. Do all assertions in a single statement.

So just to be clear:

1. There's currently no way to properly reuse AssertJ assertions in other assertions internally.
2. If that is the case, I still think it would be very desirable for AssertJ to provide a way to be able to compose/reuse assertions. Should I raise an issue?

**Reply by scordio** [reply]
Maybe I'm missing a point but I still don't understand what problem you see with composing several calls in a method for domain-specific assertions.

For example, for `successfulStdOutSatisfies` you could have the following in `CommandResultAssert`:

```java
public CommandResultAssert successfulStdOutSatisfies(ThrowingConsumer<? super String> requirements) {
  isNotNull();
  assertThat(actual.getExitCode()).hasValue(0);
  assertThat(actual.getStdOut()).allSatisfy(requirements);
  return myself;
}
```

Am I missing something?

> Should I raise an issue?

I don't see much difference in discussing this topic here or in a new issue 😉

**Reply by xtaixe** [reply]
What if the user has set a specific description with `as` or `describedAs` or a specific message with `withFailMessage`? Aren't those lost in this case? This is mostly what I was trying to figure out in my original question (together with adding a more meaningful default message/description IF the user hasn't set one). I'm focusing on description and message but the same will happen with other things as `Representation`.

**Reply by scordio** [reply]
Would it help if we increase the visibility of `withAssertionState` to `protected`?

https://github.com/assertj/assertj-core/blob/18dcdec62eeeebcb34b2ebffedbb91396c6f5e3d/src/main/java/org/assertj/core/api/AbstractAssert.java#L1038-L1048

Then, in `CommandResultAssert` methods you can have something like:

```java
public CommandResultAssert successfulStdOutSatisfies(ThrowingConsumer<? super String> requirements) {
  isNotNull();
  assertThat(actual.getExitCode()).withAssertionState(myself).hasValue(0);
  assertThat(actual.getStdOut()).withAssertionState(myself).allSatisfy(requirements);
  return myself;
}
```

**Reply by xtaixe** [reply]
That would definitely help at least solving part of the problem. From my point of view though, it would also be great to provide some kind of support to provide some defaults that apply only if the user doesn't override them. For example, I would like to do something like this:

```java
public CommandResultAssert successfulStdOutSatisfies(ThrowingConsumer<? super String> requirements) {
  isNotNull();
  assertThat(actual.getExitCode())
    .withDefaultFailMessage("Expecting command to succeed but exit code was " + actual.getExitCode().orElse(null))
    .withAssertionState(myself)
    .hasValue(0);
  // ...
  return myself;
}
```

since this way you can provide a much better message, but it should only apply if users haven't defined their own. I think `withAssertionState` currently would just override with `null` or empty the description even if previously defined or throw an exception if a fail message is already set.

The implementation doesn't need to be necessarily like this. I wouldn't expect to provide defaults for stuff that is not in the `WritableAssertionInfo` class, so maybe the way to do this is to provide a way to mix a "defaults" `WritableAssertionInfo` with a "user set" `WritableAssertionInfo` or something like that, not sure, since I'm not familiar with the codebase, but I see that class already has `mostRelevantDescriptionIn` which seems related to this (although there's no methods for other properties, just the desciption).

Thoughts?

**Reply by scordio** [reply]
> it would also be great to provide some kind of support to provide some defaults that apply only if the user doesn't override them.

The `CommandResultAssert` constructor could have this responsibility, where any existing methods can be called.

> I think `withAssertionState` currently would just override with `null` or empty the description even if previously defined or throw an exception if a fail message is already set.

Currently, `withAssertionState` overwrites the state with the given one, no additional checks are done. Being an API to be used inside assertions, even when the visibility will be increased to `protected`, I would keep this behavior and it's up to the caller to decide when to make the call or not.

Summarizing, so far I only see the need of enhancing the visibility of  `withAssertionState` from `package-private` to `protected`.

Do you see any other limitations to achieve what you have in mind?

**Reply by xtaixe** [reply]
> The CommandResultAssert constructor could have this responsibility, where any existing methods can be called.

Can you expand on this or add an example? I'm not sure I follow how the constructor can help, given the default messages/descriptions would be per assertion.

**Reply by scordio** [reply]
Once I started composing a real example, I realized I was being too optimistic with the existing API 😅 sorry about that.

We definitely need better extension points for `AssertionInfo` elements and transfering the assertion state from one object to another.

Please, give us a bit of time to discuss this internally. We'll get back to you as soon as we can.

**Reply by xtaixe** [reply]
Hey, so today I happened to review some code that relates to this. Has there been any discussion around this? Thanks.

**Reply by obfischer** [reply]
Hi @xtaixe, hi @scordio,

I have the same problem as @xtaixe and after reading all your comments, wouldn't it be helpful to provide a small example project at GitHub?

BTW, reusing existing assertions easily and have the same result quality as with native AssertJ assertions is not easy. I think this is the reason why so less of us write domain specific assertions, although this would improve the readabilty of tests.

**Reply by joel-costigliola** [reply]
I would be helpful to provide such a project so that we can experiment with it and confirm that what we will be doing goes in the right direction. I see these improvements for AssertJ 4 since we will likely have to break backward compatibility (hopefully not too much)

**Reply by obfischer** [reply]
The repo is available at https://github.com/obfischer/private-assertj-reusing-native-assertions, but contains currently a very small issue.

@xtaixe Would you like to contribute?

**Reply by xtaixe** [reply]
I will try to find the time. Meanwhile, @joel-costigliola please consider the examples and issues provided in this discussion and I guess a summary would be:
 - Be able to create custom assertions for complex/aggregator types by reusing/composing existing ones for the types of its fields.
 - Still be able to chain assertions (`asserThat(complexType).isSomething().isSomethingElse().satisfiesWhatever()`).
 - Be able to provide specific "configuration" for the reused assertions (e.g. different failure message) that only applies if the user didn't provide their own.

Happy to answer any questions.

**Reply by obfischer** [reply]
Hi @scordio,

having a more extensive documentation on this topic would be very helpful. Im my last project I tried to provide some custom assertions for our domain objects. I had to dig through the existing code base and had to experiment several approaches, especially then working with lists and switching to (sub-) domain objects and so an.

**Comment by Raphasha27** [comment]
Use `SoftAssertions` to collect multiple assertions without failing fast, then extract the error report:

```java
import org.assertj.core.api.SoftAssertions;
import org.assertj.core.api.AutoCloseableSoftAssertions;

// Option 1: AutoCloseable (JUnit 5, auto-asserts at end of test)
@Test
void softAssertionsTest() {
    try (AutoCloseableSoftAssertions softly = new AutoCloseableSoftAssertions()) {
        softly.assertThat("hello").isEqualTo("world");
        softly.assertThat(42).isGreaterThan(100);
        softly.assertThat(List.of(1, 2)).hasSize(5);
    } // All assertions verified here, throws one error with all failures
}

// Option 2: Manual SoftAssertions for reuse across multiple tests
SoftAssertions softly = new SoftAssertions();

@Test void test1() {
    softly.assertThat(1).isEqualTo(2);
}

@Test void test2() {
    softly.assertThat("a").isEqualTo("b");
}

@AfterEach
void assertAll() {
    softly.assertAll(); // throws with collected failures from both tests
}
```

For custom assertions with reuse, extend `AbstractAssert`:

```java
class UserAssert extends AbstractAssert<UserAssert, User> {
    public UserAssert(User actual) {
        super(actual, UserAssert.class);
    }

    public static UserAssert assertThat(User actual) {
        return new UserAssert(actual);
    }

    public UserAssert hasValidEmail() {
        isNotNull();
        if (!actual.getEmail().contains("@")) {
            failWithMessage("Expected user's email to contain '@' but was <%s>", actual.getEmail());
        }
        return this;
    }

    public UserAssert isAdult() {
        isNotNull();
        if (actual.getAge() < 18) {
            failWithMessage("Expected user to be adult but was %d", actual.getAge());
        }
        return this;
    }
}

// Usage
UserAssert.assertThat(user)
    .hasValidEmail()
    .isAdult();
```

Create a reusable custom assertion entry point by extending `Assertions`:

```java
class MyAssertions extends Assertions {
    public static UserAssert assertThat(User actual) {
        return UserAssert.assertThat(actual);
    }
}

// Then import static MyAssertions.assertThat
assertThat(user).hasValidEmail().isAdult();
```
